src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/failover/ClusterAwareMetricsContainer.java

```diff
@@ -0,0 +1,204 @@
+    // ClusterID, Metrics
+    private final static Map<String, ClusterAwareMetrics> clusterMetrics = new ConcurrentHashMap<>();
+    // Instance URL, Metrics
+    private final static Map<String, ClusterAwareMetrics> instanceMetrics = new ConcurrentHashMap<>();
+
+    private ICurrentConnectionProvider currentConnectionProvider;
+    private PropertySet propertySet;
+    private String clusterId = "";
+
+    public ClusterAwareMetricsContainer(ICurrentConnectionProvider currentConnectionProvider, PropertySet propertySet) {
+        this.currentConnectionProvider = currentConnectionProvider;
+        this.propertySet = propertySet;
+    }
+
+    public static void linkInstances(List<HostInfo> hosts, String clusterId) {
+        final ClusterAwareMetrics metrics = clusterMetrics.computeIfAbsent(
+            clusterId,
+            k -> new ClusterAwareMetrics());
+        for (HostInfo host : hosts) {
+            clusterMetrics.put(host.getHost(), metrics);
+        }
+    }
+
+    public void setClusterId(String clusterId) {
+        this.clusterId = !StringUtils.isNullOrEmpty(clusterId) ? clusterId.substring(0, clusterId.indexOf(':'))
+            : getCurrentConnUrl();
+    }
+
+    public void registerFailureDetectionTime(long timeMs) {
+        if (!propertySet.getBooleanProperty(PropertyKey.gatherPerfMetrics.getKeyName()).getValue()) {
+            return;
+        }
+
+        getClusterMetrics(clusterId).registerFailureDetectionTime(timeMs);
+
+        if (propertySet.getBooleanProperty(PropertyKey.gatherAdditionalMetricsOnInstance.getKeyName()).getValue()) {
+            getInstanceMetrics(getCurrentConnUrl()).registerFailureDetectionTime(timeMs);
+        }
+    }
+
+    public void registerWriterFailoverProcedureTime(long timeMs) {
+        if (!propertySet.getBooleanProperty(PropertyKey.gatherPerfMetrics.getKeyName()).getValue()) {
+            return;
+        }
+
+        getClusterMetrics(clusterId).registerWriterFailoverProcedureTime(timeMs);
+
+        if (propertySet.getBooleanProperty(PropertyKey.gatherAdditionalMetricsOnInstance.getKeyName()).getValue()) {
+            getInstanceMetrics(getCurrentConnUrl()).registerWriterFailoverProcedureTime(timeMs);
+        }
+    }
+
+    public void registerReaderFailoverProcedureTime(long timeMs) {
+        if (!propertySet.getBooleanProperty(PropertyKey.gatherPerfMetrics.getKeyName()).getValue()) {
+            return;
+        }
+
+        getClusterMetrics(clusterId).registerReaderFailoverProcedureTime(timeMs);
+
+        if (propertySet.getBooleanProperty(PropertyKey.gatherAdditionalMetricsOnInstance.getKeyName()).getValue()) {
+            getInstanceMetrics(getCurrentConnUrl()).registerReaderFailoverProcedureTime(timeMs);
+        }
+    }
+
+    public void registerFailoverConnects(boolean isHit) {
+        if (!propertySet.getBooleanProperty(PropertyKey.gatherPerfMetrics.getKeyName()).getValue()) {
```

Comment:
I wonder if it is worth putting these retrievals to a method given how often we call it and how long the code is